xtask/src/codegen/gen_lint_completions.rs

```diff
@@ -0,0 +1,115 @@
+//! Generates descriptors structure for unstable feature from Unstable Book
+use std::{
+    fs::File,
+    io::Read,
+    path::{Path, PathBuf},
+};
+
+use quote::quote;
+use walkdir::WalkDir;
+
+use crate::{
+    codegen::{project_root, reformat, update, Mode, Result},
+    not_bash::{fs2, run},
+    run_rustfmt,
+};
+
+pub fn generate_lint_completions(mode: Mode) -> Result<()> {
+    if !Path::new("./target/rust").exists() {
+        run!("git clone https://github.com/rust-lang/rust ./target/rust")?;
```

Comment:
Looks like we use this to get the latest version of unstable book sources, right?
In this case, ` --depth=1` can be used to speed up the checkout significantly.
On my machine:

<details>
<summary>2m40s, regular</summary>

```
$ time git clone https://github.com/rust-lang/rust
Cloning into 'rust'...
remote: Enumerating objects: 5, done.
remote: Counting objects: 100% (5/5), done.
remote: Compressing objects: 100% (5/5), done.
remote: Total 1332497 (delta 0), reused 1 (delta 0), pack-reused 1332492
Receiving objects: 100% (1332497/1332497), 555.05 MiB | 4.05 MiB/s, done.
Resolving deltas: 100% (1076555/1076555), done.
Updating files: 100% (25437/25437), done.

real	2m40.415s
user	1m4.343s
sys	0m19.727s
```

</details>


vs

<details>
<summary>0m19s, smallest depth</summary>

```
$ time git clone https://github.com/rust-lang/rust --depth=1
Cloning into 'rust'...
remote: Enumerating objects: 26618, done.
remote: Counting objects: 100% (26618/26618), done.
remote: Compressing objects: 100% (24792/24792), done.
remote: Total 26618 (delta 2051), reused 8448 (delta 1225), pack-reused 0
Receiving objects: 100% (26618/26618), 15.34 MiB | 2.91 MiB/s, done.
Resolving deltas: 100% (2051/2051), done.
Updating files: 100% (25437/25437), done.

real	0m19.829s
user	0m2.128s
sys	0m3.812s
```

</details>


In general, the way we try to solve this task give me a feeling that it is a nice feature for something upstream: clippy or at least rustup.
Well, maybe some other day :)

Reply:
Note that not all nightly features have an entry in the in-tree unstable book. Before publication, unstable-book-gen is ran. It generates stub entries for any added unstable features. A better method of obtaining the unstable names is to fetch the generated docs (e.g. from [this](https://static.rust-lang.org/dist/rust-docs-nightly-i686-unknown-linux-gnu.tar.gz) link) and then list the files in the unstable book.

Also you can try depending on the tidy crate from the rustc codebase and then give it access to the rustc sources (e.g. via such a git clone, or by downloading them from e.g. https://static.rust-lang.org/dist/rustc-nightly-src.tar.gz). It exposes an API that lists all lib features. This would be far more involved though than the docs method.

Reply:
thanks @SomeoneToIgnore `depth=1` added